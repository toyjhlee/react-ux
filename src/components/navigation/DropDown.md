### DropDown

```js
import React, {useState} from 'react'
import DropDown from './DropDown'
;<DropDown label="OPEN MENU">
  <DropDown.Item
    onClick={() => {
      alert(1)
    }}
  >
    Action
  </DropDown.Item>
  <DropDown.Item>Another action</DropDown.Item>
  <DropDown.Item active={true}>BB</DropDown.Item>
</DropDown>
```

```js
<select>
  <option>1</option>
  <option>2</option>
  <option>3</option>
</select>
```
