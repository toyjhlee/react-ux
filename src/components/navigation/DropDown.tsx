import React, {ReactElement, useEffect, useState} from 'react'
import styled from 'styled-components'

const Wrap = styled.div`
  position: relative;
  display: inline-block;
`

const Button = styled.button`
  border: none;
  border: 1px solid #dcdcdc;
  &:focus {
    border: 1px solid red;
  }
`

// 이 Component 의 UX 는 다음을 참고한다 https://material-ui.com/components/menus/
// tab 으로 상위 Button 가능
// 상위 Button 에서 클릭 시 option open
// option open 시 상하 이동 키로 이동 가능
// option 에 focus 가 있을 시 enter 로 선택 가능 -> 상위 Button 에 선택된 항목 설정
// option enter, click 시 focus 는 상위 Button 으로 이동
// option 은 tab 으로 이동 못 함 // 하는 버전을 하나 더 만들자

const MenuList = styled.ul`
  border-radius: 2px;
  border: 1px solid #dcdcdc;
  position: absolute;
  z-index: 1;

  left: 0;
  top: 100%;

  background-color: #fff;

  &[aria-hidden='true'] {
    display: none;
  }

  max-width: 200px;
`

interface MenuItemProps {
  active?: boolean
}

const Item = styled.li<MenuItemProps>`
  display: block;
  padding: 0.78571429rem 1.14285714rem;
  &:hover {
    background-color: rgba(0, 0, 0, 0.05);
  }

  &[aria-selected='true'] {
    color: red;
  }

  cursor: pointer;
  white-space: nowrap;
`

// interface IItem {
//   text: string
//   value?: number
//   key?: string
// }

interface DropDownProps {
  label: string
  children: React.ReactNode
}

DropDown.defaultProps = {
  menuList: [],
}

function DropDown(props: DropDownProps) {
  const {label, children} = props
  const [isOpen, setIsOpen] = useState<boolean>(false)
  const [focusIndex, setFocusIndex] = useState<number>(-1)

  const [childLength, setChildLength] = useState<number>(
    React.Children.toArray(children).length,
  )

  useEffect(() => {
    setChildLength(React.Children.toArray(children).length)
  }, [children])

  useEffect(() => {
    if (isOpen === false) {
      setFocusIndex(-1)
    }
  }, [isOpen])

  const onKeydown = (e: React.KeyboardEvent) => {
    if (e.key === 'Tab') {
      return
    }
    e.preventDefault()
    if (e.key === 'Enter') {
      if (isOpen === false) {
        setIsOpen(true)
        return
      }

      if (focusIndex === -1 && isOpen) {
        setIsOpen(false)
        return
      }

      const child: any = React.Children.toArray(children)[focusIndex]
      setIsOpen(false)
      child.props.onClick && child.props.onClick()
    }

    if (e.key === 'ArrowUp') {
      setFocusIndex(Math.max(-1, focusIndex - 1))
    }
    if (e.key === 'ArrowDown') {
      setFocusIndex(Math.min(childLength - 1, focusIndex + 1))
    }
  }

  const contents: ReactElement[] = []
  React.Children.forEach(children, (child: ReactElement, index) => {
    contents.push(
      React.cloneElement(child, {
        contents: {child},
        key: index,
        'aria-selected': child.props.active || index === focusIndex,
        onClick: () => {
          setIsOpen(false)

          child.props.onClick && child.props.onClick()
        },
      }),
    )
  })

  return (
    <Wrap onKeyDown={e => onKeydown(e)}>
      <Button
        onClick={() => setIsOpen(!isOpen)}
        onFocus={() => {
          //   setIsOpen(true)
        }}
        onBlur={() => {
          setIsOpen(false)
        }}
      >
        {label}
      </Button>
      <MenuList aria-hidden={!isOpen} data-testid={'menulist'}>
        {contents}
      </MenuList>
    </Wrap>
  )
}

DropDown.Item = Item

export default DropDown
