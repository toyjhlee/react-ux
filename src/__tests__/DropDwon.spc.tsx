import '@testing-library/jest-dom/extend-expect'
import {render, fireEvent} from '@testing-library/react'
import React from 'react'
import renderer from 'react-test-renderer'

import DropDown from '../components/navigation/DropDown'

describe('DropDown', () => {
  let container: any
  let menulist: any
  let firstItem: any
  let activeItem: any
  let openButton: any

  let onClick: any
  beforeEach(() => {
    onClick = jest.fn()
    container = render(
      <DropDown label="OPEN MENU">
        <DropDown.Item onClick={onClick}>Action</DropDown.Item>
        <DropDown.Item>Another action</DropDown.Item>
        <DropDown.Item active={true}>Active Item</DropDown.Item>
        <DropDown.Item>Something</DropDown.Item>
      </DropDown>,
    )

    openButton = container.getByText('OPEN MENU')
    menulist = container.queryByTestId('menulist')
    firstItem = document.querySelector('li:first-child')
    activeItem = document.querySelector('li:nth-child(3)')
  })

  test('default', () => {
    expect(menulist).toHaveStyle('display: none;')
    expect(firstItem).not.toHaveStyle('color: red')
    expect(activeItem).toHaveStyle('color: red')
  })

  test('keyDown open', () => {
    openButton.focus()
    fireEvent.keyDown(openButton, {key: 'Enter'})
    expect(menulist).toHaveStyle('display: block')
    fireEvent.keyDown(openButton, {key: 'ArrowDown'})

    expect(firstItem).toHaveStyle('color: red')
    expect(activeItem).toHaveStyle('color: red')

    fireEvent.keyDown(openButton, {key: 'Enter'})

    expect(onClick).toBeCalledTimes(1)
    expect(menulist).toHaveStyle('display: none')

    fireEvent.keyDown(openButton, {key: 'Enter'})

    // 닫힌 상태에서는 onClick 이 호출되는게 아니다
    // 닫힌 상태에서는 열려야 한다
    expect(onClick).toBeCalledTimes(1)
    expect(menulist).toHaveStyle('display: block')
  })

  test('keyDown Move', () => {
    openButton.focus()
    fireEvent.keyDown(openButton, {key: 'Enter'})
    expect(menulist).toHaveStyle('display: block')

    fireEvent.keyDown(openButton, {key: 'ArrowUp'})

    fireEvent.keyDown(openButton, {key: 'ArrowDown'})
    expect(document.querySelector('li:nth-child(1)')).toHaveStyle('color: red')
    fireEvent.keyDown(openButton, {key: 'ArrowDown'})
    expect(document.querySelector('li:nth-child(2)')).toHaveStyle('color: red')
    fireEvent.keyDown(openButton, {key: 'ArrowDown'})
    expect(document.querySelector('li:nth-child(3)')).toHaveStyle('color: red')
    fireEvent.keyDown(openButton, {key: 'ArrowDown'})
    expect(document.querySelector('li:nth-child(4)')).toHaveStyle('color: red')
    fireEvent.keyDown(openButton, {key: 'ArrowDown'})
    expect(document.querySelector('li:nth-child(4)')).toHaveStyle('color: red')
  })

  test('enter', () => {
    openButton.focus()

    fireEvent.keyDown(openButton, {key: 'Enter'})
    expect(menulist).toHaveStyle('display: block')

    fireEvent.keyDown(openButton, {key: 'Enter'})
    expect(menulist).toHaveStyle('display: none')
  })

  test('tab', () => {
    openButton.focus()
    fireEvent.keyDown(openButton, {key: 'Enter'})
    expect(menulist).toHaveStyle('display: block')

    openButton.blur()
    fireEvent.keyDown(openButton, {key: 'Tab'})
    expect(menulist).toHaveStyle('display: none')
  })

  test('click open', () => {
    fireEvent.click(openButton)
    expect(menulist).toHaveStyle('display: block')
    fireEvent.click(firstItem)
    expect(onClick).toBeCalledTimes(1)
    expect(menulist).toHaveStyle('display: none')
  })
})

describe('DropDown', () => {
  test('snapshot', () => {
    const component = renderer.create(
      <DropDown label="OPEN MENU">
        <DropDown.Item onClick={() => {}}>Action</DropDown.Item>
        <DropDown.Item>Another action</DropDown.Item>
        <DropDown.Item active={true}>Active Item</DropDown.Item>
        <DropDown.Item>Something</DropDown.Item>
      </DropDown>,
    )

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
