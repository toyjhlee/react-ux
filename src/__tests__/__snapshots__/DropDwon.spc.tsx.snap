// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown 1`] = `
<div
  className="sc-bdfBwQ bIPwkc"
  onKeyDown={[Function]}
>
  <button
    className="sc-gsTCUz igZWfW"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
  >
    OPEN MENU
  </button>
  <ul
    aria-hidden={true}
    className="sc-dlfnbm QmoUQ"
    data-testid="menulist"
  >
    <li
      aria-selected={false}
      className="sc-hKgILt iFIRZH"
      onClick={[Function]}
    >
      Action
    </li>
    <li
      aria-selected={false}
      className="sc-hKgILt iFIRZH"
      onClick={[Function]}
    >
      Another action
    </li>
    <li
      aria-selected={true}
      className="sc-hKgILt iFIRZH"
      onClick={[Function]}
    >
      Active Item
    </li>
    <li
      aria-selected={false}
      className="sc-hKgILt iFIRZH"
      onClick={[Function]}
    >
      Something
    </li>
  </ul>
</div>
`;

exports[`DropDown snapshot 1`] = `
<div
  className="sc-bdfBwQ bIPwkc"
  onKeyDown={[Function]}
>
  <button
    className="sc-gsTCUz igZWfW"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
  >
    OPEN MENU
  </button>
  <ul
    aria-hidden={true}
    className="sc-dlfnbm QmoUQ"
    data-testid="menulist"
  >
    <li
      aria-selected={false}
      className="sc-hKgILt iFIRZH"
      onClick={[Function]}
    >
      Action
    </li>
    <li
      aria-selected={false}
      className="sc-hKgILt iFIRZH"
      onClick={[Function]}
    >
      Another action
    </li>
    <li
      aria-selected={true}
      className="sc-hKgILt iFIRZH"
      onClick={[Function]}
    >
      Active Item
    </li>
    <li
      aria-selected={false}
      className="sc-hKgILt iFIRZH"
      onClick={[Function]}
    >
      Something
    </li>
  </ul>
</div>
`;
